### DISCUSSION TITLE
Determining the individuals for each generation from the checkpoint pickle file

### DISCUSSION THREAD

**Discussion by fcavarretta-ualr** [BlueBrain/BluePyOpt#475] [General]
From a checkpoint pickle file, how can I determine the individuals for each generation?

**Comment by AurelienJaquier** [comment]
Hi @FrancescoCavarretta !
You can find the individuals for each generation in the `"history"` key of the checkpoint. Here is the documentation for history: https://deap.readthedocs.io/en/master/api/tools.html#history

I have also made a small code for getting all of the individuals for all generation, and also all individual in a given generation. You will just have to give it a `checkpoint_path` and the number of individuals in a generation as `offspring_size`:

```
import pickle

with open(checkpoint_path, "rb") as checkpoint_file:
    cp = pickle.load(checkpoint_file, encoding="latin1")

history = cp["history"]

# -- get all of the individuals for all the generations -- #
all_individuals = list(history.genealogy_history.values())

# -- get all of the individuals for a given generation -- #
# generation number: 0 for 1st gen, 1 for 2nd gen, etc
gen_number = 0
individuals_keys = range(gen_number * offspring_size, offspring_size * (gen_number + 1))
# + 1 because in history 1st generation starts at 1
individuals = [history.genealogy_history[key + 1] for key in individuals_keys]
```
